Add themes caching for catalog

vknaisl opened the pull request.

kburger commented on src/main/java/nl/dtls/fairdatapoint/service/metadata/catalog/CatalogMetadataService.java:
```diff
@@ -53,6 +59,22 @@ public CatalogMetadataService(@Value("${metadataProperties.catalogSpecs:}") Stri
+            Set<IRI> set = new TreeSet<>(Comparator.comparing(IRI::toString));
+            set.addAll(catalog.getThemeTaxonomys());
+            set.addAll(themes);
+            catalog.setThemeTaxonomys(new ArrayList<>(set));
```
How is the theme taxonomy serialized at the moment? As a collection of `IRI`s? I think we might need to capture some more information about it (depends on how we type it maybe).

vknaisl replied:
`dcat:themeTaxonomy <https://purl.org/example#theme>, <https://www.wikidata.org/wiki/Q27318> .`

Yes, it's a list of URI. What data do you want to add?

kburger replied:
Depends on how the serialization is done. The taxonomy instance itself needs some more info (like `skos:ConceptScheme type, and maybe an `rdfs:label`).

vknaisl replied:
Should we do it in this PR or should we solve it separately?

kburger replied:
Probably separately, if it is serialized as it is now we have some ground for discussion.

vknaisl replied:
ok

kburger commented on src/main/java/nl/dtls/fairdatapoint/config/AclConfig.java:
```diff
@@ -58,19 +59,15 @@
-    @Bean
-    public AclCache aclCache() {
-        Cache springCache = cacheManager().getCache(CACHE_NAME);
+    public AclCache aclCache(ConcurrentMapCacheManager cacheManager) {
```
Any reason to go for the specific `ConcurrentMap` implementation over the parent `CacheManager` interface for injection?

vknaisl replied:
Yes, because you can't create new caches with `CacheManager`

kburger replied:
The javadoc of the `CacheManager.getCache()` interface states:
> Note that the cache may be lazily created at runtime if the native provider supports it.

I don't see any explicit invocations of the `ConcurrentMapCacheManager` method `createConcurrentMapCache(String name)` in this PR 😕

vknaisl replied:
Because it's hidden in a `setCacheNames ` method 😃

vknaisl commented:
@kburger I added tests. Can you please look at it?

kburger commented on src/main/java/nl/dtls/fairdatapoint/database/rdf/repository/repository/RepositoryMetadataRepositoryImpl.java:
```diff
@@ -0,0 +1,49 @@
+                .stream()
+                .map(s -> VALUEFACTORY.createIRI(s.getValue("fdp").stringValue()))
+                .findFirst()
+                .orElse(null);
```
Maybe either `orElseThrow` or change the return type to `Optional<IRI>`? Up for discussion, personally I'd like to ban direct `null` usage as much as I can.

kburger commented on src/main/resources/nl/dtls/fairdatapoint/database/rdf/repository/catalog/getDatasetThemesForCatalog.sparql:
```diff
@@ -0,0 +1,6 @@
+prefix dcat: <http://www.w3.org/ns/dcat#>
+
+SELECT ?theme WHERE {
```
@rajaram5 would it make sense to have a `DISTINCT(?theme)` here? The code probably stores the themes in a `Set`, so we wouldn't need to worry about duplication. Potential performance gain if there are **a lot** of themes?

vknaisl replied:
It would be better I guess. 

But still, you have to deal with duplications when you merge themes with theme taxonomies.

rajaram5 replied:
Indeed, it is better to have `DISTINCT`

vknaisl replied:
I added DISTINCT

kburger commented on src/test/java/nl/dtls/fairdatapoint/database/rdf/repository/catalog/CatalogMetadataRepositoryTest.java:
```diff
@@ -0,0 +1,109 @@
+        when(cacheManager.getCache(DATASET_THEMES_OF_CATALOG_CACHE)).thenReturn(cache);
+        when(repository.getConnection()).thenReturn(repositoryConnection);
+        when(repositoryConnection.prepareTupleQuery(any())).thenReturn(tupleQuery);
+        when(tupleQuery.evaluate()).thenReturn(tupleQueryResult);
```
What's your opinion on using an in-memory store (using `new SailRepository(new MemoryStore())` and `initialize()` on it, 2 LOC) vs mocking all underlying components (11 LOC including the class member declarations)?

vknaisl replied:
I will try it ;-)

vknaisl commented:
@kburger please check it again

kburger commented:
I've added a comment, but I don't see other changes. Are there any specifics I need to look at?

vknaisl commented:
I just addressed the issues which we discussed. I marked them as resolved with a comment. There is no new commit; I amend the changes into the previous commit.
